### PR TITLE
Fix Binder NI to IL fallback

### PR DESCRIPTION
**Pull request by sdmaclea**
Always try NI first

### CODE REVIEW

**Comment by sdmaclea** [comment]
Fixes #11324 if S.P.C.ni.dll is renamed to S.P.C.dll as current design intends.

@janvorli @gkhanna79 @parjong PTAL

File: src/binder/assemblybinder.cpp [hunk 1]
```diff
@@ -1594,10 +1589,16 @@ namespace BINDER_SPACE
             {
                 BOOL hasHeader = TRUE;
                 IF_FAIL_GO(BinderHasNativeHeader(pNativePEImage, &hasHeader));
+
                 if (!hasHeader)
                 {
-                     pPEImage = pNativePEImage;
```
Comment (sdmaclea) [inline src/binder/assemblybinder.cpp]: Treating the `pNativePEImage` as a `pPEImage`.  The `pNativePEImage` had a flag set saying it was an trusted native image.  Releasing and rebinding is safer.

File: src/vm/peimage.inl [hunk 2]
```diff
@@ -481,8 +481,19 @@ inline BOOL PEImage::HasNativeHeader()
         return GetLoadedLayout()->HasNativeHeader();
     else
     {
-        PEImageLayoutHolder pLayout(GetLayout(PEImageLayout::LAYOUT_ANY,LAYOUT_CREATEIFNEEDED));
```
Comment (sdmaclea) [inline src/vm/peimage.inl]: On arm64 platform at least this throws if it is not a native image.  Functional native images do not throw, so treating throw as result = false seems sane.

File: src/binder/assemblybinder.cpp [hunk 3]
```diff
@@ -689,11 +689,6 @@ namespace BINDER_SPACE
         sCoreLib = sCoreLibDir;
         sCoreLib.Append(CoreLibName_IL_W);
         BOOL fExplicitBindToNativeImage = (fBindToNativeImage == true)? TRUE:FALSE;
-#if defined(FEATURE_PAL) && !defined(_TARGET_AMD64_)      
-        // Non-Amd64 platforms on non-Windows do not support generating the NI image
-        // as CoreLib.dll. For those, we will bind as IL.
-        fExplicitBindToNativeImage = FALSE;
```
Comment (sdmaclea) [inline src/binder/assemblybinder.cpp]: This disabled trying to load as NI for these platforms.

File: src/binder/assemblybinder.cpp (hunk 1 above)
Comment (gkhanna79) [inline src/binder/assemblybinder.cpp]: I agree.

File: src/vm/peimage.inl (hunk 2 above)
Comment (gkhanna79) [inline src/vm/peimage.inl]: So attempting to bind to IL version of SPC.dll would throw on Arm64? Why is that?

Comment (sdmaclea) [inline src/vm/peimage.inl]: At least part of the reason was that the `ntHeader.OptionalHeader.Magic` is `IMAGE_NT_OPTIONAL_HDR32_MAGIC` for an IL image, but is expected to be `IMAGE_NT_OPTIONAL_HDR64_MAGIC` for `MAPMapPEFile` on arm64.  https://github.com/dotnet/coreclr/blob/master/src/pal/src/map/map.cpp#L2371-L2378.  When I tried to "fix" this, it failed somewhere else.  It seemed like a bad idea to try to make IL succeed.

Comment (sdmaclea) [inline src/vm/peimage.inl]: `MAPMapPEFile` error is thrown from `PEImageLayout::Map`

Comment (gkhanna79) [inline src/vm/peimage.inl]: Short of asking the callstack at which the exception is thrown, is it thrown during the call at https://github.com/dotnet/coreclr/pull/11341/files#diff-1708de55e5220e56e33f7ddb0f465f9cR1591? If so, should we have the try/catch at that point and then proceed to bind as IL as your change enables?

Comment (sdmaclea) [inline src/vm/peimage.inl]: Yes it is thrown during the call BinderHasNativeHeader above.  We could move the try catch block there.  I placed it in PEImage::HasNativeHeader() as that seemed like the most general location to place this.  Let me know what @gkhanna79  and @janvorli prefer.

Comment (sdmaclea) [inline src/vm/peimage.inl]: I'm living in North Carolina.  I suppose local vernacular would be what y'all prefer.

**Comment by sdmaclea** [comment]
Ubuntu failure is a ```FAILED   - Interop/MarshalAPI/String/StringMarshalingTest/StringMarshalingTest.sh``` which failed last two PRtest runs (including this one)

OSX CI also looks unstable as many of the recent prtest runs have failed.

File: src/vm/peimage.inl (hunk 2 above)
Comment (gkhanna79) [inline src/vm/peimage.inl]: Thank you @sdmaclea. I would suggest to handle it closer to the intended place where we really want to handle, while keeping the other use-cases unaffected.

Even at the proposed location, I would suggest enabling the try/catch only for Arm64 so that we are intentional about where this fallback is going to work.

Comment (sdmaclea) [inline src/vm/peimage.inl]: @gkhanna79  I was assuming there were similar issues with BinderHasNativeHeader() for arm32 and the catch would resolve the arm32 issues which caused you to disable loading native image in the first place.

Comment (gkhanna79) [inline src/vm/peimage.inl]: If you can validate it, that will be great. I know that this issue does not exist on Windows so, at minimum, the try/catch needs to be for FEATURE_PAL only.

Comment (sdmaclea) [inline src/vm/peimage.inl]: @gkhanna79 I do not have an arm32 platform to verify this locally (although the arm32 CI passed with this change.). I will amend the change to be ARM64 specific.  arm32 contribs can reference this if they need similar fixes.

File: src/vm/coreassemblyspec.cpp [hunk 4]
```diff
@@ -275,8 +275,23 @@ STDAPI BinderAcquirePEImage(LPCWSTR   wszAssemblyPath,
 
 STDAPI BinderHasNativeHeader(PEImage *pPEImage, BOOL* result)
```
Comment (sdmaclea) [inline src/vm/coreassemblyspec.cpp]: @gkhanna79 This is not exactly what you asked for, but I noticed that this function was missing the TRY/CATCH block that its peers had and was always returning S_OK.  I think this is a better fix, since it make this function correctly implement its implied contract.

File: src/binder/assemblybinder.cpp [hunk 5]
```diff
@@ -1593,11 +1593,22 @@ namespace BINDER_SPACE
             if (pNativePEImage)
             {
                 BOOL hasHeader = TRUE;
-                IF_FAIL_GO(BinderHasNativeHeader(pNativePEImage, &hasHeader));
+
+                hr = BinderHasNativeHeader(pNativePEImage, &hasHeader);
+
+#if !(defined(FEATURE_PAL) && defined(_TARGET_ARM64_))
+                IF_FAIL_GO(hr);
+#endif
```
Comment (sdmaclea) [inline src/binder/assemblybinder.cpp]: I personally do not think these three lines are necessary....

Comment (gkhanna79) [inline src/binder/assemblybinder.cpp]: You can undo this change so that it is the same as before.

**Comment by sdmaclea** [comment]
@gkhanna79 Thanks.  As this is it addresses my issues.  I will amend again tomorrow to address the undo.  It will need an ifdef in `BinderHasNativeHeader`, but it should be clean.

**Comment by sdmaclea** [comment]
@gkhanna79 This fixes my needs and I think fixes the undo.

@parjong I did not attempt to reenable native image loading for arm32/x86 as I do not have access to an arm32/x86 platform.  I think this change should make it easier to reenable.  It maybe as simple as removing the `#if fExplicitBindToNativeImage = FALSE; ` in src/binder/assemblybinder.cpp.  I would make sure both S.P.C NI and S.P.C IL both work.

**Comment by gkhanna79** [comment]
Looks good to me - should I merge this @sdmaclea ?

**Comment by sdmaclea** [comment]
@gkhanna79 Please merge at your convenience
